apps/prairielearn/src/lib/server-jobs-2.ts

```diff
@@ -0,0 +1,184 @@
+import AnsiUp from 'ansi_up';
+import execa = require('execa');
+import { z } from 'zod';
+import * as Sentry from '@prairielearn/sentry';
+import { logger } from '@prairielearn/logger';
+import { loadSqlEquiv, queryAsync, queryValidatedOneRow } from '@prairielearn/postgres';
+
+import { chalk } from './chalk';
+import serverJobs = require('./server-jobs');
+import socketServer = require('./socket-server');
+
+const sql = loadSqlEquiv(__filename);
+
+interface CreateServerJobOptions {
+  courseId?: string;
+  courseInstanceId?: string;
+  courseRequestId?: string;
+  assessmentId?: string;
+  userId?: string;
+  authnUserId?: string;
+  type: string;
+  description: string;
+}
+
+interface ServerJobExecOptions {
+  cwd?: string;
+  env?: NodeJS.ProcessEnv;
+}
+
+type ServerJobContext = Omit<ServerJob, 'execute' | 'executeInBackground'>;
+type JobExecutionFunction = (context: ServerJobContext) => Promise<void>;
+
+class ServerJob {
+  public jobSequenceId: string;
+  public jobId: string;
+  private started = false;
+  private finished = false;
+  public output = '';
+
+  constructor(jobSequenceId: string, jobId: string) {
+    this.jobSequenceId = jobSequenceId;
+    this.jobId = jobId;
+  }
+
+  error(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  warn(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  info(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  verbose(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
```

Comment:
Do we want to add some colors to these various outputs to mirror what happens if you call `logger.warn()` with console output? And should we have some kind of section headers (a `header()` method?) that can both color headers and in the future perhaps provide collapsible sections like GitHub actions?

Reply:
I like the idea of colors for warnings/errors - I can add that.

For collapsible sections, I'm envisioning an API like this:

```ts
await job.group('Name of group', async () => {
  // Do things in group
  job.info('here is a log');
});
```

If we borrow [GitHub's syntax](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines) directly, that would output something like this:

```txt
::group::Name of group
here is a log
::endgroup::
```

I'm inclined to avoid implementing this for now - this will most likely require shifting the rendering of logs to the client in order to support groups inside streaming logs.

Reply:
I added red/yellow/dim colors for error/warn/verbose, respectively, in e4e2748.

Reply:
Nice! I'm on board with deferring groups to a later time.